chain/syncer.go

```diff
@@ -421,13 +345,39 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	haltOnState := func(t types.TipSetKey) bool {
+		if syncer.chainStore.HasTipSetAndState(ctx, t.String()) {
```

Comment:
@ZenGround0 nit: the store interface would be nicer if it took tipset keys directly instead of strings, abstracting the underlying serialization format and limitations.